ReactAndroid/src/main/java/com/facebook/react/modules/network/OkHttpClientProvider.java

```diff
@@ -64,21 +68,30 @@ public static OkHttpClient createClient() {
       .writeTimeout(0, TimeUnit.MILLISECONDS)
       .cookieJar(new ReactCookieJarContainer());
 
-    return enableTls12OnPreLollipop(client).build();
+    return enableTls12OnLollipopAndBelow(client).build();
   }
 
   /*
-    On Android 4.1-4.4 (API level 16 to 19) TLS 1.1 and 1.2 are
+    On Android 4.1-5.0 (API level 16 to 21) TLS 1.1 and 1.2 are
     available but not enabled by default. The following method
     enables it.
    */
-  public static OkHttpClient.Builder enableTls12OnPreLollipop(OkHttpClient.Builder client) {
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN && Build.VERSION.SDK_INT <= Build.VERSION_CODES.KITKAT) {
+  public static OkHttpClient.Builder enableTls12OnLollipopAndBelow(OkHttpClient.Builder client) {
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN && Build.VERSION.SDK_INT <= Build.VERSION_CODES.LOLLIPOP) {
       try {
-        client.sslSocketFactory(new TLSSocketFactory());
+        TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(
+                        TrustManagerFactory.getDefaultAlgorithm());
+        trustManagerFactory.init((KeyStore) null);
+        TrustManager[] trustManagers = trustManagerFactory.getTrustManagers();
+        if (trustManagers.length != 1 || !(trustManagers[0] instanceof X509TrustManager)) {
+          throw new IllegalStateException("Unexpected default trust managers:" + Arrays.toString(trustManagers));
+        }
+        X509TrustManager trustManager = (X509TrustManager) trustManagers[0];
+
+        client.sslSocketFactory(new TLSSocketFactory(trustManager), trustManager);
 
         ConnectionSpec cs = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
-                .tlsVersions(TlsVersion.TLS_1_2)
+                .tlsVersions(TlsVersion.TLS_1_0, TlsVersion.TLS_1_1, TlsVersion.TLS_1_2)
```

Comment:
Not sure we want to enable older TLS, especially 1.0. Probably better to leave it at 1.2 only.

Reply:
TLS 1.0 is still supported by Android platform and by Okhttp, so react native should not force to TLS1.2 just to fix an issue. For Lolipop+, tls 1.0 is already available, so I'd rather enable it on older devices too in order to keep version behaviour consistent. By adding all tls versions, we let the server make the decision

Reply:
Is there a way to see which protocols the Android device supports by default and add only more secure ones?

Reply:
TLS version is decided together by the client and the server. Server and client both share the list of tls versions supported and the highest version that is supported on both platforms is selected. 

Adding TLS 1.0 here is only useful if the server doesn't support TLS 1.1 and 1.2, which is possible for a development server or a qa environment.

Reply:
I have a PR #17309 that relates to this. It is important to recognize that some Android environments do not support TLS v1.0. Does this code require checking the list of supported protocols explicitly?

Reply:
At this point why have a custom `ConnectionSpec` instance. As far as I see you can just use [`ConnectionSpec.MODERN_TLS`](https://github.com/square/okhttp/blob/okhttp_36/okhttp/src/main/java/okhttp3/ConnectionSpec.java#L68-L70) as is, without creating a builder on top of it. It is already set to 

     .tlsVersions(TlsVersion.TLS_1_3, TlsVersion.TLS_1_2, TlsVersion.TLS_1_1, TlsVersion.TLS_1_0)

In addition [`ConnectionSpec.COMPATIBLE_TLS`](https://github.com/square/okhttp/blob/okhttp_36/okhttp/src/main/java/okhttp3/ConnectionSpec.java#L75-L76) is built on top of `ConnectionSpec.MODERN.TLS`.

So if I am not misunderstanding the entire `specs` list could just be

    List<ConnectionSpec> specs = new ArrayList<>();
    specs.add(ConnectionSpec.MODERN_TLS);
    specs.add(ConnectionSpec.CLEARTEXT);

Would this not be far simpler (unless I misunderstand OkHttp/TLS API)?

Reply:
We still need COMPATIBLE_TLS as a fallback on older devices but it is a good idea to use MODERN_TLS instead of a custom builder. Updated in a6d30f7